### ISSUE TITLE
Clang 11 Segmentation fault

### ISSUE THREAD

**Issue by xdevs23** [ClangBuiltLinux/linux#942]
Building Linux 5.5.11 (`-O3`, optimized for haswell), `nitrous_defconfig`: https://gitlab.com/xdevs23/linux-nitrous/
Reproduced 3 times
`uname -a`: `Linux simbox 5.5.10-arch1-1 #1 SMP PREEMPT Wed, 18 Mar 2020 08:40:35 +0000 x86_64 GNU/Linux`
`/proc/cmdline`: `BOOT_IMAGE=/boot/vmlinuz-linux root=UUID=90fcf339-9563-4688-9bde-4eabea49deb6 rw loglevel=3 quiet mds=full,nosmt va_align_addr=1`
Ninja version: `1.10.0`
OS: Arch Linux
Hardware: Intel Skylake (i5), 16 GB DDR4, swap enabled, running on bare metal

Clang version (first occurence):
```
clang version 11.0.0 (git@github.com:llvm/llvm-project.git e23d78652691f360700a08a245ee28db151844f4)
```

(`c0e1135fb0871206fd4d1c62b9de1e9de861df07` also affected)

Last version tried:
```
clang version 11.0.0 (git@github.com:llvm/llvm-project.git e8f5c072f6df444f6baea91a3533509cf4a3779f)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /mnt/building/llvm/llvm-project/llvm/build/bin
```

LLVM was compiled using following commands (as per https://clangbuiltlinux.github.io/CBL-meetup-2020-slides/nick/debugging_llvm.html#/2):
```
$ cmake .. \
-DCMAKE_BUILD_TYPE=Release \
-G Ninja \
-DCMAKE_C_COMPILER=/mnt/building/roms/xos/prebuilts/clang/host/linux-x86/clang-r353983c/bin/clang \
-DCMAKE_CXX_COMPILER=/mnt/building/roms/xos/prebuilts/clang/host/linux-x86/clang-r353983c/bin/clang++ \
-DLLVM_ENABLE_LLD=ON \
-DLLVM_ENABLE_PROJECTS="clang;lld;compiler-rt" \
-DLLVM_TARGETS_TO_BUILD="AArch64;ARM;X86"
$ ninja
```

Error message:
```
Stack dump:
0.      Program arguments: /mnt/building/llvm/llvm-project/llvm/build/bin/clang-11 -cc1 -triple x86_64-unknown-linux-gnu -S -disable-free -disable-llvm-verifier -discard-value-names -main-file-name dir.c -mrelocation-model static -mthread-model posix -fno-delete-null-pointer-checks -mllvm -warn-stack-size=2048 -mframe-pointer=none -relaxed-aliasing -fmath-errno -fno-rounding-math -masm-verbose -no-integrated-as -mconstructor-aliases -mcmodel=kernel -target-cpu haswell -target-feature +retpoline-indirect-calls -target-feature +retpoline-indirect-branches -target-feature -sse -target-feature -mmx -target-feature -sse2 -target-feature -3dnow -target-feature -avx -target-feature -x87 -target-feature +retpoline-external-thunk -disable-red-zone -dwarf-column-info -fno-split-dwarf-inlining -debugger-tuning=gdb -nostdsysteminc -nobuiltininc -resource-dir /mnt/building/llvm/llvm-project/llvm/build/lib/clang/11.0.0 -dependency-file fs/ocfs2/.dir.o.d -MT fs/ocfs2/dir.o -sys-header-deps -isystem /mnt/building/llvm/llvm-project/llvm/build/lib/clang/11.0.0/include -include ./include/linux/kconfig.h -include ./include/linux/compiler_types.h -I ./arch/x86/include -I ./arch/x86/include/generated -I ./include -I ./arch/x86/include/uapi -I ./arch/x86/include/generated/uapi -I ./include/uapi -I ./include/generated/uapi -D __KERNEL__ -D CONFIG_AS_CFI=1 -D CONFIG_AS_CFI_SIGNAL_FRAME=1 -D CONFIG_AS_CFI_SECTIONS=1 -D CONFIG_AS_SSSE3=1 -D CONFIG_AS_AVX=1 -D CONFIG_AS_AVX2=1 -D CONFIG_AS_AVX512=1 -D CONFIG_AS_SHA1_NI=1 -D CONFIG_AS_SHA256_NI=1 -D CC_USING_FENTRY -I fs/ocfs2 -D MODULE -D KBUILD_BASENAME="dir" -D KBUILD_MODNAME="ocfs2" -fmacro-prefix-map=./= -O3 -Wall -Wundef -Werror=strict-prototypes -Wno-trigraphs -Werror=implicit-function-declaration -Werror=implicit-int -Wno-format-security -Werror=unknown-warning-option -Wno-sign-compare -Wno-frame-address -Wno-address-of-packed-member -Wno-format-invalid-specifier -Wno-gnu -Wno-tautological-compare -Wno-unused-const-variable -Wdeclaration-after-statement -Wvla -Wno-pointer-sign -Werror=date-time -Werror=incompatible-pointer-types -Wno-initializer-overrides -Wno-format -Wno-sign-compare -Wno-format-zero-length -std=gnu89 -fno-dwarf-directory-asm -fdebug-compilation-dir /home/simao/linux-nitrous/src/linux-nitrous -ferror-limit 19 -fmessage-length 0 -pg -mfentry -fwrapv -stack-protector 2 -mstack-alignment=8 -fcf-protection=none -fwchar-type=short -fno-signed-wchar -fgnuc-version=4.2.1 -fobjc-runtime=gcc -fdiagnostics-show-option -fcolor-diagnostics -vectorize-loops -vectorize-slp -o /tmp/dir-49967b.s -x c fs/ocfs2/dir.c
1.      <eof> parser at end of file
2.      Code generation
3.      Running pass 'Function Pass Manager' on module 'fs/ocfs2/dir.c'.
4.      Running pass 'Branch Probability Basic Block Placement' on function '@ocfs2_prepare_dir_for_insert'
 #0 0x00000000026e7b24 PrintStackTraceSignalHandler(void*) (/mnt/building/llvm/llvm-project/llvm/build/bin/clang-11+0x26e7b24)
 #1 0x00000000026e57de llvm::sys::RunSignalHandlers() (/mnt/building/llvm/llvm-project/llvm/build/bin/clang-11+0x26e57de)
 #2 0x00000000026e7f1c SignalHandler(int) (/mnt/building/llvm/llvm-project/llvm/build/bin/clang-11+0x26e7f1c)
 #3 0x00007f1a00d5e800 __restore_rt (/usr/lib/libpthread.so.0+0x14800)
 #4 0x000000000202314b (anonymous namespace)::MachineBlockPlacement::buildChain(llvm::MachineBasicBlock const*, (anonymous namespace)::BlockChain&, llvm::SmallSetVector<llvm::MachineBasicBlock const*, 16u>*) (/mnt/building/llvm/llvm-project/llvm/build/bin/clang-11+0x202314b)
 #5 0x000000000201d246 (anonymous namespace)::MachineBlockPlacement::buildCFGChains() (/mnt/building/llvm/llvm-project/llvm/build/bin/clang-11+0x201d246)
 #6 0x000000000201c3ef (anonymous namespace)::MachineBlockPlacement::runOnMachineFunction(llvm::MachineFunction&) (/mnt/building/llvm/llvm-project/llvm/build/bin/clang-11+0x201c3ef)
 #7 0x0000000001e02a44 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/mnt/building/llvm/llvm-project/llvm/build/bin/clang-11+0x1e02a44)
 #8 0x000000000215cde9 llvm::FPPassManager::runOnFunction(llvm::Function&) (/mnt/building/llvm/llvm-project/llvm/build/bin/clang-11+0x215cde9)
 #9 0x000000000215d053 llvm::FPPassManager::runOnModule(llvm::Module&) (/mnt/building/llvm/llvm-project/llvm/build/bin/clang-11+0x215d053)
#10 0x000000000215d5ef llvm::legacy::PassManagerImpl::run(llvm::Module&) (/mnt/building/llvm/llvm-project/llvm/build/bin/clang-11+0x215d5ef)
#11 0x000000000288980c clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::DataLayout const&, llvm::Module*, clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream> >) (/mnt/building/llvm/llvm-project/llvm/build/bin/clang-11+0x288980c)
#12 0x00000000030253eb clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (/mnt/building/llvm/llvm-project/llvm/build/bin/clang-11+0x30253eb)
#13 0x0000000003c617f3 clang::ParseAST(clang::Sema&, bool, bool) (/mnt/building/llvm/llvm-project/llvm/build/bin/clang-11+0x3c617f3)
#14 0x0000000002f818a1 clang::FrontendAction::Execute() (/mnt/building/llvm/llvm-project/llvm/build/bin/clang-11+0x2f818a1)
#15 0x0000000002ededaf clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/mnt/building/llvm/llvm-project/llvm/build/bin/clang-11+0x2ededaf)
#16 0x000000000301fa7b clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/mnt/building/llvm/llvm-project/llvm/build/bin/clang-11+0x301fa7b)
#17 0x00000000015bfd52 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/mnt/building/llvm/llvm-project/llvm/build/bin/clang-11+0x15bfd52)
#18 0x00000000015be05c ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&) (/mnt/building/llvm/llvm-project/llvm/build/bin/clang-11+0x15be05c)
#19 0x00000000015bde14 main (/mnt/building/llvm/llvm-project/llvm/build/bin/clang-11+0x15bde14)
#20 0x00007f1a007c5023 __libc_start_main (/usr/lib/libc.so.6+0x27023)
#21 0x00000000015bb02e _start (/mnt/building/llvm/llvm-project/llvm/build/bin/clang-11+0x15bb02e)
  CC [M]  sound/soc/codecs/pcm186x.o
clang-11: error: unable to execute command: Segmentation fault (core dumped)
clang-11: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 11.0.0 (git@github.com:llvm/llvm-project.git e8f5c072f6df444f6baea91a3533509cf4a3779f)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /mnt/building/llvm/llvm-project/llvm/build/bin
clang-11: note: diagnostic msg: PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
clang-11: note: diagnostic msg:
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang-11: note: diagnostic msg: /tmp/dir-e8d2b0.c
clang-11: note: diagnostic msg: /tmp/dir-e8d2b0.sh
clang-11: note: diagnostic msg:

********************
```

Log: 
First occurence: [clang11-crash.txt](https://github.com/ClangBuiltLinux/linux/files/4365056/clang11-crash.txt)
Third occurence: 
[clang11-crash-2.txt](https://github.com/ClangBuiltLinux/linux/files/4365168/clang11-crash-2.txt)

Files to attach:

[dir-e8d2b0.c.txt](https://github.com/ClangBuiltLinux/linux/files/4365173/dir-e8d2b0.c.txt)
[dir-e8d2b0.sh.txt](https://github.com/ClangBuiltLinux/linux/files/4365174/dir-e8d2b0.sh.txt)

**Comment by nathanchance** [comment]
creduce test case:

```shell
#!/bin/bash
# -O3 is most critical for reproducing this bug, the other flags are ones that seemed to be needed as well but I did not verify carefully

gcc -O3 -fno-delete-null-pointer-checks -fstack-protector-strong -c -o /dev/null dir.i || exit ${?}

/home/nathan/src/tc-build/build/llvm/stage1/bin/clang -O3 -no-integrated-as -fno-delete-null-pointer-checks -fstack-protector-strong -c -o /dev/null dir.i |& grep "Branch Probability Basic Block Placement"
```

Result:

```
e, m, n, at, au, p;
struct j {
  long d;
  short k;
  char l
} * o;
struct g *bf;
b() {
  _Bool a;
  asm goto("" : : : : c);
  return a;
c:;
}
struct g {
  char h
} ab() {
  if (e)
    i();
}
an(struct j *q, struct g *r) {
  char *ap = 0;
  if (__builtin_expect(q->k % 4, 0))
    ap = "";
  else if (__builtin_expect(r->h > m, 0))
    ap = "directory entry across blocks";
  if (__builtin_expect(ap, 0))
    aq(ap);
  return ap == 0;
}
ar(int q, struct j *r) {
  if (q != r->l)
    return 0;
  if (!r->d)
    return 0;
  return memcmp(n, r, q);
}
av(struct j *q) {
  if (at)
    return 1;
  au = q;
  if (q->l)
    return 1;
  return 0;
}
static bg(int q, struct g **r) {
  int bk = bl();
  if (bk)
    goto bail;
  while (1) {
    if (bf->h) {
      if (bm())
        goto bail;
      bk = 0;
      o = bf->h;
    }
    if (!an(o, bf))
      goto bail;
    if (ar(q, o))
      goto bail;
    if (av(o)) {
      *r = bo();
      goto bail;
    }
    e = o->k;
  }
bail:
  ab();
  return bk;
}
bt(int q, struct g r, int s, struct bv *t) {
  int ba;
  struct g *ac = b();
  if (!s)
    goto bd;
  if (f())
    goto bd;
  if (p)
    ba = 0;
  else
    ba = bg(s, &ac);
  bx(q, r, t);
  struct g by = *ac;
  i(by);
bd:
  return ba;
}
```

Reducing in LLVM IR:

```
$ export PATH=/home/nathan/src/tc-build/build/llvm/stage1/bin:${PATH}

$ clang -O3 -no-integrated-as -march=haswell -fno-delete-null-pointer-checks -fstack-protector-strong -emit-llvm -S -o dir.ll dir.i

$ llc -O3 dir.ll
Stack dump:
0.      Program arguments: llc -O3 dir.ll
1.      Running pass 'Function Pass Manager' on module 'dir.ll'.
2.      Running pass 'Branch Probability Basic Block Placement' on function '@bt'
 #0 0x0000000001899ed4 PrintStackTraceSignalHandler(void*) (/home/nathan/src/tc-build/build/llvm/stage1/bin/llc+0x1899ed4)
 #1 0x0000000001897bfe llvm::sys::RunSignalHandlers() (/home/nathan/src/tc-build/build/llvm/stage1/bin/llc+0x1897bfe)
 #2 0x000000000189a2cc SignalHandler(int) (/home/nathan/src/tc-build/build/llvm/stage1/bin/llc+0x189a2cc)
 #3 0x00007f52a4fab730 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x12730)
 #4 0x0000000000ed321b (anonymous namespace)::MachineBlockPlacement::buildChain(llvm::MachineBasicBlock const*, (anonymous namespace)::BlockChain&, llvm::SmallSetVector<llvm::MachineBasicBlock const*, 16u>*) (/home/nathan/src/tc-build/build/llvm/stage1/bin/llc+0xed321b)
 #5 0x0000000000ecd2ee (anonymous namespace)::MachineBlockPlacement::buildCFGChains() (/home/nathan/src/tc-build/build/llvm/stage1/bin/llc+0xecd2ee)
 #6 0x0000000000ecc4af (anonymous namespace)::MachineBlockPlacement::runOnMachineFunction(llvm::MachineFunction&) (/home/nathan/src/tc-build/build/llvm/stage1/bin/llc+0xecc4af)
 #7 0x0000000000f06cb4 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/home/nathan/src/tc-build/build/llvm/stage1/bin/llc+0xf06cb4)
 #8 0x0000000001253223 llvm::FPPassManager::runOnFunction(llvm::Function&) (/home/nathan/src/tc-build/build/llvm/stage1/bin/llc+0x1253223)
 #9 0x0000000001253533 llvm::FPPassManager::runOnModule(llvm::Module&) (/home/nathan/src/tc-build/build/llvm/stage1/bin/llc+0x1253533)
#10 0x0000000001253adf llvm::legacy::PassManagerImpl::run(llvm::Module&) (/home/nathan/src/tc-build/build/llvm/stage1/bin/llc+0x1253adf)
#11 0x00000000009869d5 main (/home/nathan/src/tc-build/build/llvm/stage1/bin/llc+0x9869d5)
#12 0x00007f52a48ac09b __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2409b)
#13 0x000000000098402a _start (/home/nathan/src/tc-build/build/llvm/stage1/bin/llc+0x98402a)
zsh: segmentation fault  llc -O3 dir.ll

$ cat bugpoint.sh
#!/bin/bash

! /home/nathan/src/tc-build/build/llvm/stage1/bin/llc -O3 "${1}" |& grep "Branch Probability Basic Block Placement"

$ bugpoint -compile-custom -compile-command=./bugpoint.sh dir.ll
...

$ llvm-dis bugpoint-reduced-simplified.bc

$ llc -O3 bugpoint-reduced-simplified.ll
Stack dump:
0.      Program arguments: llc -O3 bugpoint-reduced-simplified.ll
1.      Running pass 'Function Pass Manager' on module 'bugpoint-reduced-simplified.ll'.
2.      Running pass 'Branch Probability Basic Block Placement' on function '@bt'
 #0 0x0000000001899ed4 PrintStackTraceSignalHandler(void*) (/home/nathan/src/tc-build/build/llvm/stage1/bin/llc+0x1899ed4)
 #1 0x0000000001897bfe llvm::sys::RunSignalHandlers() (/home/nathan/src/tc-build/build/llvm/stage1/bin/llc+0x1897bfe)
 #2 0x000000000189a2cc SignalHandler(int) (/home/nathan/src/tc-build/build/llvm/stage1/bin/llc+0x189a2cc)
 #3 0x00007ff1c78e7730 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x12730)
 #4 0x0000000000ed321b (anonymous namespace)::MachineBlockPlacement::buildChain(llvm::MachineBasicBlock const*, (anonymous namespace)::BlockChain&, llvm::SmallSetVector<llvm::MachineBasicBlock const*, 16u>*) (/home/nathan/src/tc-build/build/llvm/stage1/bin/llc+0xed321b)
 #5 0x0000000000ecd2ee (anonymous namespace)::MachineBlockPlacement::buildCFGChains() (/home/nathan/src/tc-build/build/llvm/stage1/bin/llc+0xecd2ee)
 #6 0x0000000000ecc4af (anonymous namespace)::MachineBlockPlacement::runOnMachineFunction(llvm::MachineFunction&) (/home/nathan/src/tc-build/build/llvm/stage1/bin/llc+0xecc4af)
 #7 0x0000000000f06cb4 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/home/nathan/src/tc-build/build/llvm/stage1/bin/llc+0xf06cb4)
 #8 0x0000000001253223 llvm::FPPassManager::runOnFunction(llvm::Function&) (/home/nathan/src/tc-build/build/llvm/stage1/bin/llc+0x1253223)
 #9 0x0000000001253533 llvm::FPPassManager::runOnModule(llvm::Module&) (/home/nathan/src/tc-build/build/llvm/stage1/bin/llc+0x1253533)
#10 0x0000000001253adf llvm::legacy::PassManagerImpl::run(llvm::Module&) (/home/nathan/src/tc-build/build/llvm/stage1/bin/llc+0x1253adf)
#11 0x00000000009869d5 main (/home/nathan/src/tc-build/build/llvm/stage1/bin/llc+0x9869d5)
#12 0x00007ff1c71e809b __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2409b)
#13 0x000000000098402a _start (/home/nathan/src/tc-build/build/llvm/stage1/bin/llc+0x98402a)
zsh: segmentation fault  llc -O3 bugpoint-reduced-simplified.ll
```

`bugpoint-reduced-simplified.ll`:

```
; ModuleID = 'bugpoint-reduced-simplified.bc'
source_filename = "dir.i"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

%struct.j.0.3.6.192.195.222.234.267 = type { i64, i16, i8 }
%struct.g.1.4.7.193.196.223.235.268 = type { i8 }
%struct.bv.2.5.8.194.197.224.236.269 = type opaque

@e = external dso_local local_unnamed_addr global i32, align 4
@.str = external dso_local unnamed_addr constant [1 x i8], align 1
@m = external dso_local local_unnamed_addr global i32, align 4
@.str.1 = external dso_local unnamed_addr constant [30 x i8], align 1
@n = external dso_local local_unnamed_addr global i32, align 4
@at = external dso_local local_unnamed_addr global i32, align 4
@au = external dso_local local_unnamed_addr global i32, align 4
@p = external dso_local local_unnamed_addr global i32, align 4
@o = external dso_local local_unnamed_addr global %struct.j.0.3.6.192.195.222.234.267*, align 8
@bf = external dso_local local_unnamed_addr global %struct.g.1.4.7.193.196.223.235.268*, align 8

declare dso_local i32 @b() #0

declare dso_local i8 @ab() local_unnamed_addr #0

declare dso_local i32 @i(...) local_unnamed_addr #0

declare dso_local i32 @an(%struct.j.0.3.6.192.195.222.234.267*, %struct.g.1.4.7.193.196.223.235.268*) local_unnamed_addr #0

declare dso_local i32 @aq(...) local_unnamed_addr #0

declare dso_local i32 @ar(i32, %struct.j.0.3.6.192.195.222.234.267*) local_unnamed_addr #0

declare dso_local i32 @memcmp(i8*, i8*, i64) local_unnamed_addr #0

declare dso_local i32 @av(%struct.j.0.3.6.192.195.222.234.267*) local_unnamed_addr #0

define dso_local i32 @bt(i32 %0, i8 %1, i32 %2, %struct.bv.2.5.8.194.197.224.236.269* %3) local_unnamed_addr #0 {
  callbr void asm sideeffect "", "X,~{dirflag},~{fpsr},~{flags}"(i8* blockaddress(@bt, %5)) #2
          to label %6 [label %5]

5:                                                ; preds = %4
  br label %6

6:                                                ; preds = %5, %4
  br i1 undef, label %70, label %7

7:                                                ; preds = %6
  %8 = tail call i32 (...) @f() #2
  %9 = icmp eq i32 %8, 0
  br i1 %9, label %10, label %70

10:                                               ; preds = %7
  %11 = icmp eq i32 undef, 0
  br i1 %11, label %12, label %63

12:                                               ; preds = %10
  %13 = tail call i32 (...) @bl() #2
  br i1 undef, label %14, label %58

14:                                               ; preds = %12
  %15 = load %struct.g.1.4.7.193.196.223.235.268*, %struct.g.1.4.7.193.196.223.235.268** @bf, align 8
  %16 = getelementptr inbounds %struct.g.1.4.7.193.196.223.235.268, %struct.g.1.4.7.193.196.223.235.268* %15, i64 0, i32 0
  %17 = load i8, i8* %16, align 1
  br label %18

18:                                               ; preds = %56, %14
  %19 = phi %struct.g.1.4.7.193.196.223.235.268* [ %15, %14 ], [ %28, %56 ]
  %20 = phi i8 [ %17, %14 ], [ %35, %56 ]
  %21 = icmp eq i8 %20, 0
  br i1 %21, label %22, label %23

22:                                               ; preds = %18
  br label %27

23:                                               ; preds = %18
  %24 = icmp eq i32 undef, 0
  br i1 %24, label %25, label %58

25:                                               ; preds = %23
  %26 = inttoptr i64 0 to %struct.j.0.3.6.192.195.222.234.267*
  br label %27

27:                                               ; preds = %25, %22
  %28 = phi %struct.g.1.4.7.193.196.223.235.268* [ %19, %22 ], [ null, %25 ]
  %29 = phi %struct.j.0.3.6.192.195.222.234.267* [ undef, %22 ], [ %26, %25 ]
  %30 = load i16, i16* undef, align 8
  %31 = and i16 %30, 3
  %32 = icmp eq i16 %31, 0
  br i1 %32, label %33, label %37, !prof !1, !misexpect !2

33:                                               ; preds = %27
  %34 = getelementptr inbounds %struct.g.1.4.7.193.196.223.235.268, %struct.g.1.4.7.193.196.223.235.268* %28, i64 0, i32 0
  %35 = load i8, i8* %34, align 1
  %36 = icmp slt i32 0, undef
  br i1 %36, label %37, label %39

37:                                               ; preds = %33, %27
  %38 = tail call i32 (i8*, ...) bitcast (i32 (...)* @aq to i32 (i8*, ...)*)(i8* nonnull undef) #2
  br label %58

39:                                               ; preds = %33
  %40 = load i8, i8* undef, align 2
  %41 = icmp eq i32 0, %2
  br i1 %41, label %42, label %48

42:                                               ; preds = %39
  %43 = icmp eq i64 undef, 0
  br i1 %43, label %48, label %44

44:                                               ; preds = %42
  %45 = bitcast %struct.j.0.3.6.192.195.222.234.267* %29 to i8*
  %46 = tail call i32 @bcmp(i8* undef, i8* %45, i64 undef) #2
  %47 = icmp eq i32 %46, 0
  br i1 %47, label %48, label %58

48:                                               ; preds = %44, %42, %39
  %49 = icmp eq i32 undef, 0
  br i1 %49, label %50, label %54

50:                                               ; preds = %48
  %51 = ptrtoint %struct.j.0.3.6.192.195.222.234.267* %29 to i64
  %52 = trunc i64 %51 to i32
  store i32 %52, i32* @au, align 4
  %53 = icmp eq i8 %40, 0
  br i1 %53, label %56, label %54

54:                                               ; preds = %50, %48
  %55 = inttoptr i64 undef to %struct.g.1.4.7.193.196.223.235.268*
  br label %58

56:                                               ; preds = %50
  %57 = sext i16 %30 to i32
  store i32 %57, i32* @e, align 4
  br label %18

58:                                               ; preds = %54, %44, %37, %23, %12
  %59 = phi %struct.g.1.4.7.193.196.223.235.268* [ null, %37 ], [ %55, %54 ], [ null, %12 ], [ null, %23 ], [ null, %44 ]
  %60 = phi i32 [ 0, %37 ], [ 0, %54 ], [ %13, %12 ], [ 0, %23 ], [ 0, %44 ]
  br i1 undef, label %63, label %61

61:                                               ; preds = %58
  %62 = tail call i32 (...) @i() #2
  br label %63

63:                                               ; preds = %61, %58, %10
  %64 = phi %struct.g.1.4.7.193.196.223.235.268* [ null, %10 ], [ %59, %58 ], [ %59, %61 ]
  %65 = phi i32 [ 0, %10 ], [ %60, %58 ], [ %60, %61 ]
  %66 = tail call i32 (i32, i8, %struct.bv.2.5.8.194.197.224.236.269*, ...) bitcast (i32 (...)* @bx to i32 (i32, i8, %struct.bv.2.5.8.194.197.224.236.269*, ...)*)(i32 %0, i8 %1, %struct.bv.2.5.8.194.197.224.236.269* %3) #2
  %67 = getelementptr inbounds %struct.g.1.4.7.193.196.223.235.268, %struct.g.1.4.7.193.196.223.235.268* %64, i64 0, i32 0
  %68 = load i8, i8* %67, align 1
  %69 = tail call i32 (i8, ...) bitcast (i32 (...)* @i to i32 (i8, ...)*)(i8 %68) #2
  br label %70

70:                                               ; preds = %63, %7, %6
  %71 = phi i32 [ undef, %7 ], [ %65, %63 ], [ undef, %6 ]
  ret i32 %71
}

declare dso_local i32 @f(...) local_unnamed_addr #0

declare dso_local i32 @bx(...) local_unnamed_addr #0

declare dso_local i32 @bl(...) local_unnamed_addr #0

declare dso_local i32 @bm(...) local_unnamed_addr #0

declare dso_local i32 @bo(...) local_unnamed_addr #0

; Function Attrs: readonly
declare i32 @bcmp(i8*, i8*, i64) local_unnamed_addr #1

attributes #0 = { "use-soft-float"="false" }
attributes #1 = { readonly }
attributes #2 = { nounwind }

!llvm.ident = !{!0}

!0 = !{!"ClangBuiltLinux clang version 11.0.0 (git://github.com/llvm/llvm-project 2ec59a0a40f4ec02e6b2dbe5f12261959c191aa9)"}
!1 = !{!"branch_weights", i32 2000, i32 1}
!2 = !{!"misexpect", i64 1, i64 2000, i64 1}
```

Thankfully this is _not_ an LLVM 10 regression:

```
git bisect good
6aebf0ee56e52afad3887b4230d7ed1beaf0bede is the first bad commit
commit 6aebf0ee56e52afad3887b4230d7ed1beaf0bede
Author: Bill Wendling <isanbard@gmail.com>
Date:   Wed Mar 11 17:07:02 2020 -0700

    Specify branch probabilities for callbr dests

    Summary:
    callbr's indirect branches aren't expected to be taken, so reduce their
    probabilities to 0 while increasing the default destination to 1. This
    allows some code improvements through block placement.

    Reviewers: nickdesaulniers

    Subscribers: hiraditya, llvm-commits

    Tags: #llvm

    Differential Revision: https://reviews.llvm.org/D72656

 .../CodeGen/SelectionDAG/SelectionDAGBuilder.cpp   |  4 ++--
 llvm/test/CodeGen/AArch64/callbr-asm-label.ll      | 12 +++++------
 llvm/test/CodeGen/AArch64/callbr-asm-obj-file.ll   |  4 ++--
 llvm/test/CodeGen/X86/callbr-asm-branch-folding.ll | 23 +++++++++++-----------
 llvm/test/CodeGen/X86/callbr-asm.ll                | 17 ++++++++--------
 5 files changed, 29 insertions(+), 31 deletions(-)

$ git bisect log
# bad: [0cc124c186a5211ae5a734fe7708d61b5a150bc2] [llvm-objdump][test] Improve PowerPC branch offset tests
# good: [5852475e2c049ce29dcb1f0da3ac33035f8c9156] Bump the trunk major version to 11
git bisect start 'origin/master' 'llvmorg-11-init'
# good: [60fea2713d3f37d70383aacaa75f61344cc3234a] AMDGPU/GlobalISel: Improve 16-bit bswap
git bisect good 60fea2713d3f37d70383aacaa75f61344cc3234a
# good: [31e03317633909c50ead53edf8a19b60698075cc] [ORC] Skip ST_File symbols in MaterializationUnit interfaces / resolution.
git bisect good 31e03317633909c50ead53edf8a19b60698075cc
# bad: [263c4a3c75a46996bec1f23264874b7f1334426a] Fix compiler warning when compiling without asserts
git bisect bad 263c4a3c75a46996bec1f23264874b7f1334426a
# good: [b28ed9cec8dd7225164eb8c0884aa463654ef3fc] [clang-format] cleanup from D75517
git bisect good b28ed9cec8dd7225164eb8c0884aa463654ef3fc
# good: [a46dba24fa35ab52e9a1bbaa52666bcc37859927] [AMDGPU] Extend macro fusion for ADDC and SUBB to SUBBREV
git bisect good a46dba24fa35ab52e9a1bbaa52666bcc37859927
# bad: [df90a15b1ac938559a8c3af12126559c1e1e9558] [lldb] Clear all settings during a test's setUp
git bisect bad df90a15b1ac938559a8c3af12126559c1e1e9558
# good: [ab69cd0779c529519eb7d26e0fa1b8dfb505f838] [X86] Support intrinsic _mm_cldemote
git bisect good ab69cd0779c529519eb7d26e0fa1b8dfb505f838
# bad: [4327a9b46b46d587816f765c619838ea3e01cd19] [AMDGPU] Use progbits type for .AMDGPU.disasm section
git bisect bad 4327a9b46b46d587816f765c619838ea3e01cd19
# bad: [c4d23d8854840294bf49c524f93e2be85a401f00] Add a missing include to clang unit tests
git bisect bad c4d23d8854840294bf49c524f93e2be85a401f00
# good: [c700e0317c25f3f397a8ba368752c4960f4ab975] [JITLink] Read symbol linkage from the correct field.
git bisect good c700e0317c25f3f397a8ba368752c4960f4ab975
# bad: [b47c9f535c8a0fffeb7634a82e3901d416915938] [libc] Add initial assert definition
git bisect bad b47c9f535c8a0fffeb7634a82e3901d416915938
# bad: [6aebf0ee56e52afad3887b4230d7ed1beaf0bede] Specify branch probabilities for callbr dests
git bisect bad 6aebf0ee56e52afad3887b4230d7ed1beaf0bede
# good: [c4cbc5806218bf4232523771805580fa40b83244] [NFC][PowerPC] Add a new MIR file te test ppc-early-ret pass
git bisect good c4cbc5806218bf4232523771805580fa40b83244
# first bad commit: [6aebf0ee56e52afad3887b4230d7ed1beaf0bede] Specify branch probabilities for callbr dests
```

cc @gwelymernans

**Comment by xdevs23** [comment]
Thanks for taking the time to reproduce and document this.

I'd like to add that Clang 10 (release/10.x) works fine. It was compiled the same way as Clang 11 and I'm only getting this on 11, not on 10.

**Comment by nickdesaulniers** [comment]
> 4.      Running pass 'Branch Probability Basic Block Placement' on function '@ocfs2_prepare_dir_for_insert'

cc @gwelymernans

**Comment by gwelymernans** [comment]
Strange...I'll take a look at it.

**Comment by gwelymernans** [comment]
It seems like a bug that was exposed by the patch. MBBs are merged and one of the deleted blocks appears to be accessed afterwards.

**Comment by xdevs23** [comment]
It's still happening. It didn't happen with LTO enabled.

[clang-11-crash.txt](https://github.com/ClangBuiltLinux/linux/files/4606129/clang-11-crash.txt)
[tmp__dir-22f304.sh.txt](https://github.com/ClangBuiltLinux/linux/files/4606131/tmp__dir-22f304.sh.txt)
[tmp__dir-22f304.c.txt](https://github.com/ClangBuiltLinux/linux/files/4606132/tmp__dir-22f304.c.txt)

**Comment by xdevs23** [comment]
Patches are different but it works when compiling 5.7.2 using clang 11

```
clang version 11.0.0 (git@github.com:llvm/llvm-project.git 4707bc217780895207a9dc7467495a92eb0106a5)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /mnt/building/llvm/llvm-project/llvm/build/bin
```

I also ran the script and the source file above and it worked fine.

```
sh dir-22f304.sh
CC=clang HOSTCC=clang PATH="/mnt/building/llvm/llvm-project/llvm/build/bin:$PATH" sh dir-22f304.sh
```

Both commands worked (it wouldn't make a difference, though)

**Comment by nickdesaulniers** [comment]
Ok, thanks for confirming there's no longer an issue w/ ToT.  We might be able to bisect and find the fix, and try to backport that to the 10.1 release.  If you checkout `origin/release/10.x` and can still repro, then we can try to find which commit (back on the master branch) fixed this by bisecting clang itself.

**Comment by xdevs23** [comment]
It does not affect Clang 10, all Clang 10 versions I have ever tried have not caused this specific issue.
